docs: translate Extracting State Logic into a Reducer (#544)
reactjs/id.react.dev: src/content/learn/extracting-state-logic-into-a-reducer.md

---

> title: Mengekstraksi Logika State ke Sebuah Reducer

Reviewer: ```suggestion
title: Mengekstraksi Logika State ke Reducer
```

Minor fix to title.

---

<Intro>

Komponen dengan banyak pembaruan _state_ yang tersebar di banyak _event handlers_ bisa menjadi sangat membingungkan. Untuk kasus seperti ini, Anda dapat menggabungkan semua logika pembaruan _state_ di luar komponen Anda dalam satu fungsi, yang disebut sebagai _reducer_.

</Intro>

<YouWillLearn>

> - Apa itu fungsi _reducer_

Reviewer 2: Always use `*` to italicized

- Bagaimana cara untuk migrasi dari fungsi `useState` menjadi `useReducer`   
- Kapan menggunakan fungsi _reducer_
- Bagaimana cara menulis fungsi _reducer_ dengan baik

</YouWillLearn>

## Mengkonsolidasikan logika state menggunakan reducer {/*consolidate-state-logic-with-a-reducer*/}

> Saat komponen-komponen Anda semakin kompleks, hal ini mengakibatkan berbagai cara memperbaharui state komponen dalam kode menjadi sulit untuk dilihat secara sekilas. Contoh, komponent `TaskApp` dibawah menyimpan senarai `tasks` dalam _state_ dan menggunakan tiga _handler_ untuk menabahkan, menghapus dan mengubah `tasks`:

Reviewer 2: Bukan "memperbaharui" tapi "memperbarui"

Reviewer 2: Typo di beberapa tempat, serta slaah penggunaan "di"
```suggestion
Saat komponen-komponen Anda semakin kompleks, hal ini mengakibatkan berbagai cara memperbarui state komponen dalam kode menjadi sulit untuk dilihat secara sekilas. Contoh, komponen `TaskApp` di bawah menyimpan senarai `tasks` dalam _state_ dan menggunakan tiga _handler_ untuk menambahkan, menghapus dan mengubah `tasks`:
```

> Saat komponen-komponen Anda semakin kompleks, hal ini mengakibatkan berbagai cara memperbaharui *state* komponen dalam kode menjadi sulit untuk dilihat secara sekilas. Contoh, komponen `TaskApp` di bawah menyimpan senarai `tasks` dalam *state* dan menggunakan tiga *handler* untuk menabahkan, menghapus dan mengubah `tasks`:

Reviewer 2: Memperbarui bukan memperbaharui

Reviewer: ```suggestion
Saat komponen-komponen Anda semakin kompleks, hal ini mengakibatkan berbagai cara memperbarui *state* komponen dalam kode menjadi sulit untuk dilihat secara sekilas. Contoh, komponen `TaskApp` di bawah menyimpan senarai `tasks` dalam *state* dan menggunakan tiga *handler* untuk menambahkan, menghapus dan mengubah `tasks`:
```

</Sandpack>

> Setiap _event handler_ pada komponen `TaskApp` akan memangil fungsi `setTask` untuk melakukan pembaharuan _state_. Saat komponen ini bertumbuh semakin besar, logika _state_ akan semakin rumit. Untuk mengurangi kompleksitas dan menyimpan semua logika Anda di satu tempat yang mudah diakses. Anda dapat memindahkan logika _state_ tersebut ke sebuah fungsi di luar komponen Anda, yang **disebut _"reducer"_.**

Reviewer 2: ```suggestion
Setiap _event handler_ pada komponen `TaskApp` akan memangil fungsi `setTask` untuk melakukan pembaruan _state_. Saat komponen ini bertumbuh semakin besar, logika _state_ akan semakin rumit. Untuk mengurangi kompleksitas dan menyimpan semua logika Anda di satu tempat yang mudah diakses. Anda dapat memindahkan logika _state_ tersebut ke sebuah fungsi di luar komponen Anda, yang **disebut _"reducer"_.**
```

_Reducer_ merupakan sebuah alternatif untuk menangani _state_. Anda dapat migrasi dari `useState` ke `useReducer` dalam tiga langkah:

> 1. **Pindah** dari memanggil fungsi penetap _state_ menjadi aksi _dispatch_.

Reviewer 2: "Dispatching Actions" ini apa ya yang enak? @resir014 @zainfathoni @resir014 @mhaidarhanif ?

```suggestion
1. **Pindah** dari menyetel _state_ menjadi aksi _dispatch_.
```

Reviewer 3: Udah diganti ya sama commit baru. Tapi ini bisa lebih enak:

```suggestion
1. **Pindah** dari menyetel *state* menjadi menugaskan *action* (*dispatching actions*).
```

Reviewer 2: +1 from me 🚀

> 1. **Pindah** dari memanggil fungsi penetap *state* menjadi aksi *dispatch*.
> 2. **Tulis** fungsi *reducer*.
> 3. **Gunakan** *reducer* pada komponen Anda.

Reviewer: ```suggestion
1. **Pindah** dari menyetel *state* menjadi men-*dispatch* *action*
2. **Tulis** fungsi *reducer*.
3. **Gunakan** *reducer* pada komponen Anda.
```

"aksi dispatch" sounds like a noun, where in this case it's supposed to be a verb, so "men-*dispatch*" sounds better.

And...well, *actions* and *dispatch* are also React*-specific pattern right? I guess we can leave it untranslated?

(*and yes, some other libraries also follow the same concept, but in general it's a very specific technical term to web development)

@mazipan @mhaidarhanif @zainfathoni

> ### Langkah 1: Pindah dari memanggil fungsi penetap state menjadi aksi dispatch {/*step-1-move-from-setting-state-to-dispatching-actions*/}

Reviewer 2: Lihat komentar sebelumnya, samain biar konsisten.

Reviewer: ```suggestion
### Langkah 1: Pindah dari menyetel state menjadi men-dispatch action {/*step-1-move-from-setting-state-to-dispatching-actions*/}
```

Idem.

Reviewer 3: Ini untuk italic:

```suggestion
### Langkah 1: Pindah dari menyetel *state* menjadi men-*dispatch* *action* {/*step-1-move-from-setting-state-to-dispatching-actions*/}
```

Reviewer: @mhaidarhanif No need to italicise for headings.

> _Event handler_ Anda sekarang menentukan apa _yang harus dilakukan_ dengan memanggil menetapkan _state_:

Reviewer 2: Memanggil menetapkan, dua kata kerja beriringan, bikin bingung.

> *Event handler* Anda sekarang menentukan apa *yang harus dilakukan* dengan memanggil fungsi penetap *state*:

Reviewer 3: ```suggestion
*Event handler* Anda sekarang menentukan apa *yang harus dilakukan* dengan menyetel *state*:
```

Tidak ada kata asal untuk `memanggil fungsi`, `setting` cukup `menyetel`

Reviewer 3: @thaddeuscleo Please fix and resolve this to be approved.

}
```

Contoh diatas merupakan objek JavaScript biasa. Anda dapat menentukan objek apapun yang akan dimasukkan ke dalamnya, tetapi umumnya harus berisi sebuah objek yang setidaknya berisi informasi *Apa yang terjadi*. (Anda akan menambahkan fungsi `dispatch` itu sendiri di langkah selanjutnya.)

<Note>

Objek aksi bisa memiliki berbagai macam bentuk.

Secara konvensi, umumnya objek aksi diberikan sebuah properti `type` bertipe *string* yang menjelaskan aksi apa yang telah terjadi. Dan menambahkan informasi tambahan pada properti lainnya. Properti `type` ini khusus untuk komponen, maka untuk contoh ini nilai `'added'` atau `'added_task'` akan baik-baik saja diberikan pada properti `type`. Tentukan nama yang dapat menjelaskan aksi apa yang terjadi!

```js
dispatch({
  // khusus untuk komponen
  type: 'what_happened',
  // properti lain disini
});
```

</Note>

### Langkah 2: Tulis fungsi reducer {/*step-2-write-a-reducer-function*/}

Fungsi *reducer* merupakan tempat dimana Anda meletakkan logika *state*. Fungsi *reducer* menerima dua argumen yaitu *state* sekarang dan object aksi, dan mengembalikan *state* berikutnya:

```js
function yourReducer(state, action) {

>   // mengembalikan state berikutnya untuk di tetapkan oleh React

Reviewer: ```suggestion
  // mengembalikan state berikutnya untuk ditetapkan oleh React
```

di tetapkan -> ditetapkan

  }
}
```

> Karena fungsi *reducer* menerima *state* (`tasks`) sebagai sebuah argument, **Anda dapat mendeklarasikan fungsi *reducer* di luar komponen Anda.** Dengan ini kode yang Anda tulis tingkat indentasinya akan berkurang dan akan lebih mudah di baca.

Reviewer: ```suggestion
Karena fungsi *reducer* menerima *state* (`tasks`) sebagai sebuah argument, **Anda dapat mendeklarasikan fungsi *reducer* di luar komponen Anda.** Dengan ini kode yang Anda tulis tingkat indentasinya akan berkurang dan akan lebih mudah dibaca.
```

di baca -> dibaca
